### ISSUE TITLE
Make X.509 CRT, CSR and CRL structs opaque

### ISSUE THREAD

**Issue by davidhorstmann-arm** [Mbed-TLS/mbedtls#9548]
Make the members of the major X.509 structures private, replacing struct member accesses with getters and setters.

This will allow flexibility in the implementation of certificate parsing. For example, it would enable us to reduce memory usage by parsing on demand rather than pre-parsing the entire certificate.

Each type of X.509 struct except CRLs has two variants, one for reading and one for writing:
* `mbedtls_x509_crt` and `mbedtls_x509write_cert`
* `mbedtls_x509_csr` and `mbedtls_x509write_csr`
* `mbedtls_x509_crl` and (no writing variant for CRLs)

The structs for reading should have only getter functions and the writing structs should have only setters.

Open questions:
* For complex sub-structures, should we make these opaque as well to give more flexibility to the implementation? For example we could add accessors for the struct's submembers to the larger struct, or provide accessors for the sub-struct as well.
* How should we deal with lists of structs within the larger structs, e.g. `mbedtls_x509_crl_entry` and `mbedtls_x509_name`? We could:
   * Design the API so as to avoid the allocation of long linked-lists of values, which would both reduce memory usage and reduce the likelihood of heap corruption bugs.
   * Keep the current approach and have the user deal with a large linked list. This is simpler for us and closer to the current API.

**Comment by gilles-peskine-arm** [comment]
> it would enable us to reduce memory usage by parsing on demand rather than pre-parsing the entire certificate.

If we want to do that, then we have to make the getters take a `mbedtls_x509_crt*` parameter (for example) and not `mbedtls_x509_crt const*`, since the first getter call will need to modify the certificate. That's going to be problematic. We may also need to document locking behavior.

**Comment by gilles-peskine-arm** [comment]
One of the reasons we kept many members of the X.509 structures public is that they're large substructures, so a getter needs to either make a copy or give temporary ownership, and a setter needs to either make a copy or transfer ownership. One of the strengths of the classic Mbed TLS APIs is that they avoid difficulties around memory ownership: everything that goes through a function is copied. We shouldn't lose this property, but we also need to be careful with memory usage.

The chained list fields are particularly problematic in this respect. If you want to iterate through a list, the library should really not copy the tail of the list at each iteration.

**Comment by hanno-becker** [comment]
You may want to have a look at the very old https://github.com/Mbed-TLS/mbedtls/pull/2478 as well, where I thought hard about how to reduce memory usage through on-demand parsing. Not to say this is what you want these days, but some of the considerations in there may be useful.

**Comment by davidhorstmann-arm** [comment]
> The chained list fields are particularly problematic in this respect. If you want to iterate through a list, the library should really not copy the tail of the list at each iteration.

I agree, and linked lists seem generally like a pain for the user - they have to take account of freeing them and it's not always clear whether and how they need to be free'd. We've had at least one [memory leak fix](https://github.com/Mbed-TLS/mbedtls/pull/6391) as a result of this ambiguity.

Would a stateful iteration-based API be a good solution to this? It would seem to make memory ownership cleaner. Something that could be used like so:
```c
void print_name(mbedtls_x509_crt const *crt)
{
    mbedtls_x509_issuer_reset(crt);
    while ((mbedtls_x509_name *name = mbedtls_x509_issuer_get_next(crt)) != NULL) {
        print_oid(name->oid);
        print_val(name->val);
    }
}
```
Real API may differ etc.
